lib/shelley/src/Cardano/Wallet/Shelley/Compatibility/Ledger.hs

```diff
@@ -0,0 +1,243 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- |
+-- Copyright: © 2020 IOHK
+-- License: Apache-2.0
+--
+-- Exposes a wallet-friendly interface to types and functions exported by the
+-- ledger specification.
+--
+module Cardano.Wallet.Shelley.Compatibility.Ledger
+    (
+      -- * Exported ledger functions
+      computeMinimumAdaQuantity
+
+      -- * Conversions from wallet types to ledger specification types
+    , toLedgerCoin
+    , toLedgerTokenBundle
+    , toLedgerTokenPolicyId
+    , toLedgerTokenName
+    , toLedgerTokenQuantity
+
+      -- * Conversions from ledger specification types to wallet types
+    , toWalletCoin
+    , toWalletTokenBundle
+    , toWalletTokenPolicyId
+    , toWalletTokenName
+    , toWalletTokenQuantity
+
+      -- * Roundtrip conversion between wallet types and ledger specification
+      --   types
+    , Convert (..)
+
+    ) where
+
+import Prelude
+
+import Cardano.Crypto.Hash
+    ( hashFromBytes, hashToBytes )
+import Cardano.Wallet.Primitive.Types.Coin
+    ( Coin (..) )
+import Cardano.Wallet.Primitive.Types.Hash
+    ( Hash (..) )
+import Cardano.Wallet.Primitive.Types.TokenBundle
+    ( TokenBundle )
+import Cardano.Wallet.Primitive.Types.TokenPolicy
+    ( TokenName (..), TokenPolicyId (..) )
+import Cardano.Wallet.Primitive.Types.TokenQuantity
+    ( TokenQuantity (..) )
+import Data.Function
+    ( (&) )
+import Data.Generics.Internal.VL.Lens
+    ( view )
+import Data.Generics.Labels
+    ()
+import Data.Word
+    ( Word64 )
+import Fmt
+    ( pretty )
+import GHC.Stack
+    ( HasCallStack )
+import Ouroboros.Consensus.Shelley.Protocol.Crypto
+    ( StandardCrypto )
+
+import qualified Cardano.Ledger.Mary.Value as Ledger
+import qualified Cardano.Ledger.ShelleyMA.Rules.Utxo as Ledger
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Cardano.Wallet.Primitive.Types.TokenMap as TokenMap
+import qualified Data.Map.Strict as Map
+import qualified Data.Map.Strict.NonEmptyMap as NonEmptyMap
+import qualified Shelley.Spec.Ledger.API as Ledger
+
+--------------------------------------------------------------------------------
+-- Exported functions
+--------------------------------------------------------------------------------
+
+-- | Uses the ledger specification to compute the minimum required ada quantity
+--   for a token bundle.
+--
+computeMinimumAdaQuantity
+    :: Coin
+    -- ^ The absolute minimum ada quantity specified by the protocol.
+    -> TokenBundle
+    -- ^ The token bundle to evaluate.
```

Comment:
The function we've been using in the rest of the code works in principle from a `TokenMap` (since we do not yet know the coin value). I think we can adjust the rest of the codebase to work off `TokenBundle`, but if doable, I'd rather fix it here.

Reply:
Hi @KtorZ . I'll push a patch to this effect very shortly. 👍🏻

Reply:
@KtorZ I've just pushed 39dfd7c1c6309a57010add65fa127630c656c4b1, which changes the type of the public function to:

```hs
-- | Uses the ledger specification to compute the minimum required ada quantity
--   for a token map.
--
computeMinimumAdaQuantity
    :: Coin
    -- ^ The absolute minimum ada quantity specified by the protocol.
    -> TokenMap
    -- ^ The token map to evaluate.
    -> Coin
    -- ^ The minimum ada quantity for the given token map.
```

The old function is kept around as an internal function. We do need to keep that around however, as:

- it's a simple wrapper around the ledger specification's `scaledMinDeposit` function, which expects to be supplied with an ada quantity
- we need to test that `scaledMinDeposit` remains constant w.r.t. the ada quantity.